src/main/scala/com/fulcrumgenomics/vcf/JointVariantContextIterator.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.vcf
+
+import htsjdk.samtools.SAMSequenceDictionary
+import htsjdk.variant.variantcontext.{VariantContext, VariantContextComparator}
+
+/**
+  * Iterates over multiple variant context iterators such that we return a list of contexts for the union of sites
+  * across the iterators.  If samples is given, we subset each variant context to just that sample.
+  */
```

Comment:
The use of sampleNames seems to me to be driven by a very specific use case that is a lot less general than the iterator itself (i.e. I want a single sample per iterator).  I think I would tend to exclude that from this class and either:
a) Just have the caller map the iterators ahead of time `(iters.zipWith(samples).map { case (i, s) => i. subContextFromSample(s) }`
b) Provide an apply method in a companion object that does the mapping of the iterators before providing them to the `JointVariantContextIterator`